File: wc-clone/wc-go/wc.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"os"
	"strings"
)

func defaultCount(filePath string) (int, int, int, error) {
    var lineCount, wordCount, byteCount int
    var err error

    if filePath == "-" {
        scanner := bufio.NewScanner(os.Stdin)
        for scanner.Scan() {
            line := scanner.Text()
            lineCount++
            wordCount += len(strings.Fields(line))
            byteCount += len(scanner.Bytes())
        }
        if err := scanner.Err(); err != nil {
            return 0, 0, 0, err
        }
    } else {
        lineCount, err = countLines(filePath)
        if err != nil {
            return 0, 0, 0, err
        } 

        wordCount, err = countWords(filePath)
        if err != nil {
            return 0, 0, 0, err
        } 

        byteCount, err = countBytes(filePath)
        if err != nil {
            return 0, 0, 0, err
        } 
    }

    return lineCount, wordCount, byteCount, nil
}

func countChars(filePath string) (int, error) {
    var data []byte
    var err error
    if filePath == "-" {
        data, err = io.ReadAll(os.Stdin)
    } else {
        data, err = os.ReadFile(filePath)
    } 
    if err != nil {
        return 0, err
    } 

    count := 0
    for range string(data) {
        count++
    } 

    return count, nil
}

func countWords(filePath string) (int, error) {
    var scanner *bufio.Scanner
    if filePath == "-" {
        scanner = bufio.NewScanner(os.Stdin)
    } else {
        file, err := os.Open(filePath)
        if err != nil {
            return 0, err
        } 
        defer file.Close()
        scanner = bufio.NewScanner(file)
    }

    scanner.Split(bufio.ScanWords)
    count := 0 
    for scanner.Scan() {
        count++
    } 
    return count, nil
} 

func countBytes(filePath string) (int, error) {
    var data []byte
    var err error
    if filePath == "-" {
        data, err = io.ReadAll(os.Stdin)
    } else {
        data, err = os.ReadFile(filePath)
    }
    if err != nil {
        return 0, err
    }
    return len(data), nil
} 

func countLines(filePath string) (int, error) {
    var scanner *bufio.Scanner
    if filePath == "-" {
        scanner = bufio.NewScanner(os.Stdin)
    } else {
        file, err := os.Open(filePath)
        if err != nil {
            return 0, err
        }
        defer file.Close()
        scanner = bufio.NewScanner(file)
    }
    count := 0
    for scanner.Scan() {
        count++
    }
    return count, nil
} 

func main() {
    countBytesFlag := flag.Bool("c", false, "Count bytes")
    countLinesFlag := flag.Bool("l", false, "Count lines")
    countWordsFlag := flag.Bool("w", false, "Count words")
    countCharsFlag := flag.Bool("m", false, "Count characters")
    flag.Parse()

    var filePath string
    if flag.NArg() == 0 {
        filePath = "-"
    } else {
        filePath = flag.Arg(0)
    } 

    if !*countBytesFlag && !*countLinesFlag && !*countWordsFlag && !*countCharsFlag {
        lineCount, wordCount, byteCount, err := defaultCount(filePath)
        if err != nil {
            fmt.Println("Error: Could not read input")
            os.Exit(1)
        } 
        fmt.Printf("%d %d %d\n", lineCount, wordCount, byteCount)
        return
    }

    if *countCharsFlag {
        count, err := countChars(filePath)
        if err != nil {
            fmt.Printf("Error: Could not open file '%s'\n", filePath)
            os.Exit(1)
        } 
        fmt.Printf("%d %s\n", count, filePath)
    } else if *countBytesFlag {
        count, err := countBytes(filePath)
        if err != nil {
            fmt.Printf("Error: Could not open file '%s'\n", filePath)
            os.Exit(1)
        }
        fmt.Printf("%d %s\n", count, filePath)
    } else if *countLinesFlag {
        count, err := countLines(filePath)
        if err != nil {
            fmt.Printf("Error: Could not open file '%s'\n", filePath)
            os.Exit(1)
        } 
        fmt.Printf("%d %s\n", count, filePath)
    } else if *countWordsFlag {
        count, err := countWords(filePath)
        if err != nil {
            fmt.Printf("Error: Could not open file '%s'\n", filePath)
            os.Exit(1)
        } 
        fmt.Printf("%d %s\n", count, filePath)
    } else {
        fmt.Println("Error: Missing -c, -l, -m,  or -w flag")
        os.Exit(1)
    }
} 

File: wc-clone/wc-elixir/lib/wc.ex
defmodule Wc do
  def main(args) do
    {opts, file_names, _} =
      OptionParser.parse(
        args,
        switches: [bytes: :boolean, lines: :boolean, words: :boolean, chars: :boolean],
        aliases: [c: :bytes, l: :lines, w: :words, m: :chars]
      )

    case {opts, file_names} do
      {[], []} ->
        process_input(:stdio, opts)

      {opts, []} ->
        process_input(:stdio, opts)

      {[], [file_name]} -> 
        case default_count(file_name) do
          {:ok, [line_count, word_count, byte_count]} ->
            IO.puts("#{line_count} #{word_count} #{byte_count} #{file_name}")
          {:error, _} ->
            IO.puts(:stderr, "Error: Could not open file '#{file_name}'")
            System.halt(1)
        end

      {opts, [file_name]} ->
        cond do
          opts[:bytes] ->
            case Wc.count_bytes(file_name) do
              {:ok, count} ->
                IO.puts("#{count} #{file_name}")
              {:error, _} ->
                IO.puts(:stderr, "Error: Could not open file '#{file_name}'")
                System.halt(1)
            end
          opts[:lines] ->
            case Wc.count_lines(file_name) do
              {:ok, count} ->
                IO.puts("#{count} #{file_name}")
              {:error, _} ->
                IO.puts(:stderr, "Error: Could not open file '#{file_name}'")
                System.halt(1)
            end
          opts[:words] ->
            case Wc.count_words(file_name) do
              {:ok, count} ->
                IO.puts("#{count} #{file_name}")
              {:error, _} ->
                IO.puts(:stderr, "Error: Could not open file '#{file_name}'")
                System.halt(1)
            end
          opts[:chars] ->
            case Wc.count_chars(file_name) do
              {:ok, count} ->
                IO.puts("#{count} #{file_name}")
              {:error, _} ->
                IO.puts(:stderr, "Error: Could not open file '#{file_name}'")
                System.halt(1)
            end
          true ->
            IO.puts(:stderr, "Usage: wc [-c] [-l] <file>")
            System.halt(1)
        end

      {_, _} ->
        IO.puts(:stderr, "Usage: wc [-c] [-l] <file>")
        System.halt(1)
    end
  end

  defp default_count(input) do
    case input do 
      :stdio -> 
        content = IO.read(:stdio, :all)
        line_count = content |> String.split("\n", trim: true) |> length()
        word_count = content |> String.split() |> length()
        byte_count = byte_size(content)
        {:ok, [line_count, word_count, byte_count]}
      file_name -> 
        case File.read(file_name) do 
          {:ok, content} -> 
            line_count = content |> String.split("\n", trim: true) |> length()
            word_count = content |> String.split() |> length()
            byte_count = byte_size(content)
            {:ok, [line_count, word_count, byte_count]}
          {:error, _ } ->
            {:error, :file_not_found}
        end
    end
  end

  def count_bytes(file_name) do
    case File.read(file_name) do
      {:ok, content} -> {:ok, byte_size(content)}
      {:error, _} -> {:error, :file_not_found}
    end
  end

  def count_lines(file_name) do
    case File.read(file_name) do
      {:ok, content} ->
        {:ok, content |> String.trim() |> String.split("\n") |> length()}
      {:error, _} ->
        {:error, :file_not_found}
    end
  end

  def count_words(file_name) do
    case File.read(file_name) do
      {:ok, content} ->
        {:ok, content |> String.split() |> length()}
      {:error, _} ->
        {:error, :file_not_found}
    end
  end

  def count_chars(file_name) do
    case File.read(file_name) do
      {:ok, content} -> {:ok, String.length(content)}
      {:error, _} -> {:error, :file_not_found}
    end
  end

  defp process_input(:stdio, opts) do
    input = IO.read(:stdio, :all)

    result =
      cond do
        opts[:bytes] -> {:ok, byte_size(input)}
        opts[:lines] -> {:ok, input |> String.split("\n") |> length()}
        opts[:words] -> {:ok, input |> String.split() |> length()}
        opts[:chars] -> {:ok, String.length(input)}
        true ->
          line_count = input |> String.trim() |> String.split("\n") |> length()
          word_count = input |> String.trim() |> String.split() |> length()
          byte_count = byte_size(input)
          {:ok, [line_count, word_count, byte_count]}
      end

    case result do
      {:ok, [line_count, word_count, byte_count]} ->
        IO.puts("#{line_count} #{word_count} #{byte_count}")
      {:ok, counts} ->
        case counts do
          [line_count, word_count, byte_count] ->
            IO.puts("#{line_count} #{word_count} #{byte_count}")
          count ->
            IO.puts(count)
        end
      {:error, _} ->
        System.halt(1)
    end
  end

end
